### ISSUE TITLE
UTCTime instance and ISO8601 format w/o separators

### ISSUE THREAD

**Issue by dzhus** [haskell/aeson#386]
```
λ> decode "\"2014-02-01T14:12:37Z\"" :: Maybe UTCTime
Just 2014-02-01 14:12:37 UTC
λ> decode "\"20140201T141237Z\"" :: Maybe UTCTime
Nothing
```

Shouldn't FromJSON instance for UTCTime also support separator-less format as described by ISO8601?

**Comment by bergmark** [comment]
I'm up for discussing this.

* I don't think we currently promise to handle ISO8601 and JSON itself doesn't specify a date format, but this doesn't mean we can't be more lenient in the future.
* As for prior art we currently have the `DotNetTime` format  as a separate newtype with instance instead of combining it with `UTCTime` parsing

The advantage of a newtype is that the parser is more straight-forward. It also means that allowing the new format is opt-in which could be good or bad...  e.g. to combine `UTCTime` and `DotNetTime` parsing you currently need to use `parseJSON x <|> (fromDotNetTime <$> parseJSON x)`

**Comment by hvr** [comment]
> I don't think we currently promise to handle ISO8601 and JSON itself doesn't specify a date format, but this doesn't mean we can't be more lenient in the future.

fwiw, the idea that resulted in considering  ISO8601 encoding as default date/time serialisation for JSON was Ecma-262 from which JSON is derived. See also #1

However, in #197 we already set the precedent for making Aeson a bit more lenient, as long as we remain sound (e.g. there must still be an explicit timezone designation if you want to parse something into a UTCTime or otherwise zoned datetime)

**Comment by dzhus** [comment]
Actually, after being bit by `ToJSON UTCTime` instance not being ECMA-262-compliant 

(
```haskell
λ> now <- getCurrentTime
λ> let truncateMillis utcTime = utcTime{utctDayTime = fromInteger $ floor $ utctDayTime utcTime}
λ> truncateMillis now
2016-05-27 13:15:27 UTC
λ> encode $ truncateMillis now
"\"2016-05-27T13:15:27.000000000000Z\""
```
)

I am no longer sure if making `FromJSON` parser more lenient is the best idea either. However, in non-ideal world it's probably worth having a lax reader with a strict printer.

**Comment by phadej** [comment]
We don't support full ISO8601 when parsing. Rather we support a (conservative extension of) RFC3339 (which is almost a subset of ISO8601, allowing ` ` instead of `T`).

Secondly, we cannot really support dates without separators, it will be hard to parse. Things like `100000101` which is `10000-10-10`. ISO8601 doesn't really say anything concrete about years with over 5 digits (nor negative).

There is https://www.iso.org/obp/ui/#iso:std:iso:8601:-2:ed-1:v1:en, but I have no access to that document.
